py3status/modules/hueshift.py

```diff
@@ -0,0 +1,177 @@
+
+    def post_config_hook(self):
+        hueshift_commands = ['^blueshift$', '^redshift$', '^sct$']
+        check_commands = [x.strip('^$') for x in hueshift_commands]
+        self.pgrep_command = ['pgrep', '|'.join(hueshift_commands)]
+        if not self.command:
+            self.command = self.py3.check_commands(check_commands)
+        elif self.command not in check_commands:
+            raise Exception(STRING_BAD_COMMAND % self.command)
+        elif not self.py3.check_commands(self.command):
+            raise Exception(STRING_NOT_INSTALLED % self.command)
+        if not self.command:
+            raise Exception(STRING_NOT_AVAILABLE)
+
+        self.hue = {
+            'blueshift': {
+                'set': lambda value: ['blueshift', '-p', '-t', str(value)],
+            },
+            'redshift': {
+                'set': lambda value: ['redshift', '-r', '-O', str(value)],
+            },
+            'sct': {
+                'set': lambda value: ['sct', str(value)],
+            },
+        }
+
+        self.default = 6500
+        self.minimum = max(self.minimum, 1000)
+        self.maximum = min(self.maximum, 25000)
+        if self.command == 'sct':
+            self.maximum = min(self.maximum, 10000)
+
+        self.temperature = self.py3.storage_get('temperature') or self.default
+        self.last_value = self.temperature
+        self._set_color_setter_boolean()
+        if not self.is_color_setter_running:
+            self._run_command('set', self.temperature)
+
+    def _run_command(self, action=None, value=None, last=None):
+        command = self.hue[self.command][action](value)
+        self.py3.command_run(command)
+        if not last:
+            self.last_value = value
+        self.temperature = value
+
+    def _set_color_setter_boolean(self):
+        # for better user experience, we check for running processes
+        # before allowing users to shift color temperature on the screen.
+        try:
+            self.py3.command_output(self.pgrep_command)
+            self.is_color_setter_running = True
+        except:
+            self.is_color_setter_running = False
+
+    def hueshift(self):
+        self.py3.threshold_get_color(self.temperature, 'hue')
+        return {
+            'cached_until': self.py3.CACHE_FOREVER,
+            'full_text': self.py3.safe_format(
+                self.format, {
+                    'temperature': self.temperature,
+                    'command': self.command,
+                    'disabled': self.is_color_setter_running,
+                })}
+
+    def kill(self):
+        self.py3.storage_set('temperature', self.last_value)
+
+    def on_click(self, event):
+        self._set_color_setter_boolean()
+        if self.is_color_setter_running:
+            return
+        value = self.temperature
+        button = event['button']
+        if button == self.button_up:
+            value = min(value + self.delta, self.maximum)
```

Comment:
Imho this is still wrong. it should look like
```
if button == self.button_up:
        self._set_hue(value + self.delta)
```
the `_set_hue()` function should check the value/remember it etc because we should do the check where it is needed rather than in an unrelated place.  It also helps with readability which is always the most important thing (in my opinion it is far more important than code that works)

Reply:
I need a bit more time if you want me to remove 'set' param because I removed 'reset' from `self.hue` not long ago or should we keep it as-is... if for some reason, somebody implement something new?  Which?

Reply:
I'm so sorry about the delay. Here you go. Removed.

Reply:
basically the test of the value shold be in the right place, imo this is not the right place for many reasons.

the current code is

* hard to read (yes I can read it but that is not the issue) it could be easy to read
* hard to read code makes bugs appear
* hard to read code makes bugs hard to find
* hard to read code is hard to review
* hard to review code does not get reviewed
* code that is not reviewed does not get merged
* code that does not get merged makes everybody sad

I'm more interested in the big picture with these things, but generally if I suggest stuff there is more reason than I'm just messing with you.  Anyhow I have things to do this evening catch you later

Reply:
Okay. Have a good evening. Put out a diff because I'm not seeing what I should be doing here. Are you talking about doing this?
```
diff --git a/py3status/modules/hueshift.py b/py3status/modules/hueshift.py
index ea6f4578..41414866 100644
--- a/py3status/modules/hueshift.py
+++ b/py3status/modules/hueshift.py
@@ -153,9 +153,11 @@ class Py3status:
         value = self.temperature
         button = event['button']
         if button == self.button_up:
-            self._set_temperature(min(value + self.delta, self.maximum))
+            value = min(value + self.delta, self.maximum)
+            self._set_temperature(value)
         elif button == self.button_down:
-            self._set_temperature(max(value - self.delta, self.minimum))
+            value = max(value - self.delta, self.minimum)
+            self._set_temperature(value)
         elif button == self.button_toggle:
             value = self.default if value != self.default else self.last_value
             self._set_temperature(value, True)
```